tools/filterdiff.py

```diff
@@ -0,0 +1,154 @@
+
+
+class FilterImageInfo:
+    """
+    Filter specific fields in the image-info that can't be compared
+    together.
+    """
+
+    def __init__(self, imi: Optional[Dict], to_remove: Optional[List]) -> None:
+        self.imi: Dict = imi if imi else {}
+        self.to_remove = to_remove if to_remove else []
+
+    def _tmpfiles_d(self):
+        """
+        Remove certain files from the tpmfiles_d that have a tendency of having
+        a non fixed content.
+        """
+        tmpfilesd = self.imi.get("tmpfiles.d")
+        if tmpfilesd:
+            for _, v in tmpfilesd.items():
+                for trm in self.to_remove:
+                    if trm in v:
+                        del v[trm]
+
+    def _lvm2(self):
+        """
+        LVM2 partitions have a UUID that is not fixed. Replace the value
+        upon comparison time.
+        """
+        partitions = self.imi.get("partitions")
+        if partitions:
+            for partition in partitions:
+                if partition.get("fstype") == "LVM2_member":
+                    partition["uuid"] = "2022-07-01-fixed-uuid"
+
+    def _iso(self):
+        """
+        For isos, the partition UUID is the date of the build. Replace
+        that with a fixed one for the comparison.
+        """
+        if ("image-format" in self.imi and "type" in self.imi["image-format"] and
+                self.imi["image-format"]["type"] == "raw"):
+            if "partitions" in self.imi:
+                for partition in self.imi["partitions"]:
+                    if "fstype" in partition:
+                        if partition["fstype"] == "iso9660":
+                            partition["uuid"] = "2022-07-01-fixed-uuid"
+
+    def apply(self) -> Dict:
+        self._iso()
+        self._lvm2()
+        self._tmpfiles_d()
+        return self.imi
+
+
+def load_item(path):
+    c = None
+    with open(path, "r") as f:
+        c = json.load(f)
+    # Supports raw image-info or DB entry
+    if "image-info" in c:
+        c = c["image-info"]
+        # Load the unwanted files from the DB
+        if "unwanted_tmpfiles_d" in c:
+            unwanted_tmpfiles_d = c["unwanted_tmpfiles_d"]
+    return c
+
+
+def search_unwanted_tmpfiles_d(path1, path2):
+    """
+    Returns the union of 
+    """
+    def search(path) -> List:
+        with open(path, "r") as f:
+            c = json.load(f)
+        # Supports raw image-info or DB entry
+        if "image-info" in c:
+            # Load the unwanted files from the DB
+            if "unwanted_tmpfiles_d" in c:
+                return c["unwanted_tmpfiles_d"]
+        return []
+    p1 = search(path1)
+    p2 = search(path2)
+    for e in p2:
+        if e not in p1:
+            p1.append(e)
+    return p1
+
+
+def diff(first, second, unwanted_tmpfiles_d, visual):
+    c1 = load_item(first)
+    c2 = load_item(second)
+    if not unwanted_tmpfiles_d:
+        unwanted_tmpfiles_d = search_unwanted_tmpfiles_d(first, second)
```

Comment:
That works only if `c1` and `c2` are DB entries. I want the diff tool to be able to diff single image-info too.

Reply:
I'm gonna mix your solution with mine though ;)

Reply:
> That works only if `c1` and `c2` are DB entries. I want the diff tool to be able to diff single image-info too.

I don't see how is that an issue. The image-info report apparently won't contain `unwanted_tmpfiles_d` key and in such case, the empty list is returned. If it is a DB entry, it will be there and its value will be returned... 🤷

Reply:
It's just that `load_item` was only returning the `image-info` component of the file. Not the entire file content. And `unwanted_tmpfiles_d` is at the same level as `image-info` in the DB entry. Now `load` returns the `image-info` and a `Set` that can be empty or not depending on what was found in the file to open.